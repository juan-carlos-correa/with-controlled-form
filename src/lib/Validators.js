export class Validators {
  constructor (value = '') {
    this.value = value
    this.errors = []
  }

  isRequired (verify, message='This value is required') {
    if (verify && !this.value) {
      this.errors.push(message);
    }

    return this;
  }

  isEmail (verify, message='This value must be a valid email. Example: name@mail.com') {
    const reg = /\S+@\S+\.\S+/;

    if (verify && !reg.test(this.value)) {
      this.errors.push(message);
    }

    return this;
  }

  isMinLength (size, message) {
    message = message || `This value must contain at least ${size} characters`

    if (!this.value || (this.value && this.value.length < size)) {
      this.errors.push(message);
    }

    return this;
  }

  isMaxLength (size, message) {
    message = message || `This value must contain a maximum of ${size} characters`

    if (!this.value || (this.value && this.value.length > size)) {
      this.errors.push(message);
    }

    return this;
  }

  isEqual (str = '', message='The values are differents') {
    if (this.value !== str) {
      this.errors.push(message);
    }

    return this;
  }

  custom (cb) {
    const { result, message } = cb(this.value);

    if (!result) {
      this.errors.push(message);
    }

    return this;
  }

  getResult () {
    const isValid = this.errors.length === 0;

    return { isValid, errors: this.errors };
  }

  static validate (values, rules = {}) {
    const result = { isValid: true, errors: {} };
    let validationOneResult = null;

    for (let value in values) {
      result.errors[value] = [];

      validationOneResult = Validators.validateOne(values[value], rules[value])

      if (!validationOneResult.isValid) {
        result.errors[value] = validationOneResult.errors;
        result.isValid = false;
      }
    }

    return result
  }

  static validateOne (value, rulesValue) {
    const result = { isValid: true, errors: [] };
    let validator = null;
    let error = null;
    let validationValue = null;

    for (let validationRule in rulesValue) {
      validator = new Validators(value);
      validationValue = rulesValue[validationRule];

      if (Array.isArray(validationValue)) {
        const value = validationValue[0]
        const message = validationValue[1]
        error = validator[validationRule](value, message).getResult().errors[0];
      } else {
        error = validator[validationRule](validationValue).getResult().errors[0];
      }

      if (!error) {
        continue;
      }

      if (result.isValid) {
        result.isValid = false;
      }

      result.errors.push(error);
    }

    return result;
  }
}
