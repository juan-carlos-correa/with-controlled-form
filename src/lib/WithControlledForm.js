import React, { Component } from 'react';
import { Validators } from './Validators';

export const WithControlledForm = (FormComponent, state = {}, formValidations = {}) => (
  class WithFormMethodsHOC extends Component {
    constructor (props) {
      super(props);

      this.state = {
        values: { ...state },
        errors: this._stateToErrors(state),
      };
    }

    _stateToErrors = (state) => {
      const errors = {};

      const keys = Object.keys(state);
      for (let i = 0; i < keys.length; i += 1) {
        errors[keys[i]] = [];
      }

      return errors;
    }

    _validateForm = (values) => {
      const result = Validators.validate(values, formValidations);
      this.setState({ errors: result.errors });
      return result.isValid;
    }

    isFormClean = () => {
      const { values } = this.state;
      const isValue = Object.values(values).filter(val => val.length > 0);
      return isValue.length === 0;
    }

    cleanForm = () => {
      const newValues = {};
      const { values } = this.state;

      for (let val in values) {
        newValues[val] = '';
      }

      this.setState({
        values: newValues,
        errors: this._stateToErrors(newValues),
      });
    }

    handleChange = (e) => {
      const { name, value } = e.target;
      const { values } = this.state;

      values[name] = value;
      this.setState({ values });
    }

    handleCheckboxChange = (e) => {
      const { name, checked } = e.target;
      const { values } = this.state;

      values[name] = checked;
      this.setState({ values });
    }

    handleBlur = (e) => {
      e.preventDefault();

      const { errors } = this.state;
      const { name, value } = e.target;
      const validations = formValidations[name];
      const result = Validators.validateOne(value, validations);
      errors[name] = !result.isValid ? result.errors[0] : [];
      this.setState({ errors });
    }

    handleSubmit = (e) => {
      e.preventDefault();
      const { values } = this.state;
      const { handleSubmit } = this.props;

      if (this._validateForm(values)) {
        handleSubmit(values);
      }
    }

    render () {
      return (
        <FormComponent
          {...this.props}
          {...this.state}
          cleanForm={this.cleanForm}
          isFormClean={this.isFormClean}
          handleChange={this.handleChange}
          handleCheckboxChange={this.handleCheckboxChange}
          handleBlur={this.handleBlur}
          handleSubmit={this.handleSubmit}
        />
      )
    }
  }
);
