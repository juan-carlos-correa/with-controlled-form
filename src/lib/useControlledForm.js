import React, { useState } from 'react'
import { Validators } from '../lib/Validators'

const _stateToErrors = (state) => {
  const errors = {};
  const names = Object.keys(state);

  for (let i = 0; i < names.length; i += 1) {
    const keyName = names[i];
    errors[keyName] = [];
  }

  return errors;
}

export const useControlledForm = (initialState, formValidations) => {
  const [values, setValues] = useState({ ...initialState });
  const [errors, setErrors] = useState(_stateToErrors(initialState));

  const handleChange = ({ target }) => {
    const { name, value } = target;

    setValues({ ...values, [name]: value });
  }

  const handleSubmit = (_handleSubmit) => {
    if (_validateForm(values)) {
      _handleSubmit(values);
    }
  }

  const _validateForm = (valuesToValidate) => {
    const result = Validators.validate(valuesToValidate, formValidations);
    setErrors({ errors: result.errors });
    return result.isValid;
  }

  return {values, errors, handleChange, handleSubmit };
}
