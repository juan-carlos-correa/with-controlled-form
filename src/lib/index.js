import WithControlledForm from './WithControlledForm';
export default WithControlledForm;
