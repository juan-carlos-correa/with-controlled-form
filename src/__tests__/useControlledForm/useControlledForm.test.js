import React from 'react';
import { mount } from 'enzyme';
import Form from '../../demo/components/Form';
import { useControlledForm } from '../../lib';

const formState = {
  email: '',
  phoneNumber: '',
};

const formValidations = {
  email: {
    isRequired: true,
    isMinLength: 3,
    isMaxLength: 30,
    isEmail: [true, 'email is not valid!'],
  },
  phoneNumber: {
    isRequired: true,
    isMinLength: 6,
    isMaxLength: 15,
  }
};

const FormWithUseControlledFormHook = ({ myHandleSubmit }) => {
  const {
    values,
    errors,
    handleChange,
    handleSubmit,
    handleBlur,
    cleanForm
   } = useControlledForm(formState, formValidations);

  return (
    <>
      <Form
        values={values}
        errors={errors}
        handleChange={handleChange}
        handleSubmit={() => handleSubmit(myHandleSubmit)}
        handleBlur={handleBlur}
      />
      <button data-test="clear-form" onClick={() => cleanForm()}>Clear form</button>
    </>
  );
}

describe('useControlledForm hook', () => {
  let handleSubmitMock = null;
  let wrapper = null;

  beforeAll(() => {
    handleSubmitMock = jest.fn();
  });

  beforeEach(() => {
    wrapper = mount(<FormWithUseControlledFormHook myHandleSubmit={handleSubmitMock} />);
  });

  afterEach(() => {
    handleSubmitMock.mockClear();
    wrapper.unmount();
  });

  describe('valid render and correct submit', () => {
    test('render without crash', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('call submit handler function properly with valid values', () => {
      const email = wrapper.find('#email');
      expect(email).toHaveLength(1);

      email.simulate('change', { target: { value: 'foo@mail.com', name: 'email' } });

      const phoneNumber = wrapper.find('#phoneNumber');
      expect(phoneNumber).toHaveLength(1);

      phoneNumber.simulate('change', { target: { value: '5512345678', name: 'phoneNumber' } });

      const form = wrapper.find('form');
      form.simulate('submit');

      expect(handleSubmitMock).toHaveBeenCalledTimes(1);
    });
  })

  describe('input validations on submit', () => {
    test('required fields', () => {
      const form = wrapper.find('form');
      form.simulate('submit');

      const emailErrorMessage = wrapper.find('#email-error-message');
      expect(emailErrorMessage.text()).toContain('This value is required');

      const phoneNumberErrorMessage = wrapper.find('#phoneNumber-error-message');
      expect(phoneNumberErrorMessage.text()).toContain('This value is required');

      expect(handleSubmitMock).toHaveBeenCalledTimes(0);
    });

    test('isMinLength', () => {
      const email = wrapper.find('#email');
      email.simulate('change', { target: { value: 'fo', name: 'email' } });

      const phoneNumber = wrapper.find('#phoneNumber');
      phoneNumber.simulate('change', { target: { value: '123', name: 'phoneNumber' } });

      const form = wrapper.find('form');
      form.simulate('submit');

      const emailErrorMessage = wrapper.find('#email-error-message');
      expect(emailErrorMessage.text()).toContain('This value must contain at least 3 characters');

      const phoneNumberErrorMessage = wrapper.find('#phoneNumber-error-message');
      expect(phoneNumberErrorMessage.text()).toContain('This value must contain at least 6 characters');

      expect(handleSubmitMock).toHaveBeenCalledTimes(0);
    });

    test('isMaxLength', () => {
      const email = wrapper.find('#email');
      email.simulate('change', { target: { value: 'foooooooooooooooooooooooo@mail.com', name: 'email' } });

      const phoneNumber = wrapper.find('#phoneNumber');
      phoneNumber.simulate('change', { target: { value: '1234567891234567', name: 'phoneNumber' } });

      const form = wrapper.find('form');
      form.simulate('submit');

      const emailErrorMessage = wrapper.find('#email-error-message');
      expect(emailErrorMessage.text()).toContain('This value must contain a maximum of 30 characters');

      const phoneNumberErrorMessage = wrapper.find('#phoneNumber-error-message');
      expect(phoneNumberErrorMessage.text()).toContain('This value must contain a maximum of 15 characters');

      expect(handleSubmitMock).toHaveBeenCalledTimes(0);
    });

    test('isEmail with custom message', () => {
      const email = wrapper.find('#email');
      email.simulate('change', { target: { value: 'foomail.com', name: 'email' } });

      const form = wrapper.find('form');
      form.simulate('submit');

      const emailErrorMessage = wrapper.find('#email-error-message');
      expect(emailErrorMessage.text()).toContain('email is not valid!');

      expect(handleSubmitMock).toHaveBeenCalledTimes(0);
    });
  });

  describe('handle blur properly', () => {
    test('with error message', () => {
      const email = wrapper.find('#email');
      expect(email).toHaveLength(1);

      email.simulate('blur', { target: { name: 'email', value: 'wrong value' } });

      const emailError = wrapper.find('div[data-test="emailError"]');
      expect(emailError.text()).toBe('email is not valid!');
    });

    test('without error message', () => {
      const email = wrapper.find('#email');
      expect(email).toHaveLength(1);

      email.simulate('blur', { target: { name: 'email', value: 'john.doe@mail.com' } });

      const emailError = wrapper.find('div[data-test="emailError"]');
      expect(emailError.text()).toBe('');
    });
  });

  describe('clean form properly', () => {
    test('clear form values and error messages', () => {
      const email = wrapper.find('#email');
      expect(email).toHaveLength(1);

      email.simulate('change', { target: { value: 'foo@mail.com', name: 'email' } });
      expect(wrapper.find('#email').instance().value).toBe('foo@mail.com');

      const phoneNumber = wrapper.find('#phoneNumber');
      expect(phoneNumber).toHaveLength(1);

      phoneNumber.simulate('change', { target: { value: '1', name: 'phoneNumber' } });
      expect(wrapper.find('#phoneNumber').instance().value).toBe('1');

      const form = wrapper.find('form');
      form.simulate('submit');

      expect(wrapper.find('div[data-test="phoneNumberError"]').text()).toBe('This value must contain at least 6 characters');

      const clearFormBtn = wrapper.find('button[data-test="clear-form"]');
      expect(clearFormBtn).toHaveLength(1);

      clearFormBtn.simulate('click');

      expect(wrapper.find('#email').instance().value).toBe('');
      expect(wrapper.find('#phoneNumber').instance().value).toBe('');
      expect(wrapper.find('div[data-test="phoneNumberError"]').text()).toBe('');
    });
  });
});
