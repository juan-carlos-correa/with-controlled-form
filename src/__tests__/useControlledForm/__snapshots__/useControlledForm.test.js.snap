// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useControlledForm hook valid render and correct submit render without crash 1`] = `
<FormWithUseControlledFormHook
  myHandleSubmit={[MockFunction]}
>
  <Form
    errors={
      Object {
        "email": Array [],
        "phoneNumber": Array [],
      }
    }
    handleBlur={[Function]}
    handleChange={[Function]}
    handleSubmit={[Function]}
    values={
      Object {
        "email": "",
        "phoneNumber": "",
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-element"
      >
        <label
          htmlFor="email"
        >
          Email:
        </label>
        <input
          className="input"
          data-test="email"
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="error-message"
          data-test="emailError"
          id="email-error-message"
        />
      </div>
      <div
        className="form-element"
      >
        <label
          htmlFor="phoneNumber"
        >
          Phone number:
        </label>
        <input
          className="input"
          data-test="phoneNumber"
          id="phoneNumber"
          name="phoneNumber"
          onBlur={[Function]}
          onChange={[Function]}
          type="tel"
          value=""
        />
        <div
          className="error-message"
          data-test="phoneNumberError"
          id="phoneNumber-error-message"
        />
      </div>
      <div
        className="text-center"
      >
        <button
          className="button"
          id="send"
          type="submit"
        >
          Send
        </button>
      </div>
    </form>
  </Form>
  <button
    data-test="clear-form"
    onClick={[Function]}
  >
    Clear form
  </button>
</FormWithUseControlledFormHook>
`;
